### ISSUE TITLE
where is auxiliary information 

### ISSUE THREAD

**Issue by ccblublu** [WongKinYiu/yolov9#196]
thanks for your excellent job,
I still have a question about multi-level auxiliary information in your paper 4.1.2
however i cannot find where it used in your code, i just find 6 heads in main branch and aux branch as shown in .yaml

**Comment by ccblublu** [comment]
@WongKinYiu

**Comment by ccblublu** [comment]
thanks for your reply, in my opinion, it is aux branch,revealed by the grey area in fig,where is the pink  area? @WongKinYiu 
![PixPin_2024-03-08_09-50-28](https://github.com/WongKinYiu/yolov9/assets/50388099/56c05d06-7523-4b62-8a80-42f49ec85c49)

**Comment by sanha9999** [comment]
https://github.com/WongKinYiu/yolov9/blob/main/models/detect/yolov9-e.yaml#L88-L105

"multi-level auxiliary branch" is pink area.

**Comment by ccblublu** [comment]
there is 3+3+3, 9 heads in fig, but just 6 heads in yaml,  which make me confused, is there any details i missed?
and as shown in yolov9-e.yaml, the relationship between main branch and aux branch is different with fig in paper,Looking forward to your relpy! @WongKinYiu @sanha9999 
![4e94bb69ec3372213726b0f69a370cde](https://github.com/WongKinYiu/yolov9/assets/50388099/563fc7b4-6ccb-42bf-af40-4a0e3c420efe)

**Comment by YoohLee** [comment]
I have this question too. I already have found the multi-level auxiliary branch in yaml files, but,  I have not find the Auxiliary Reversible Branch. And, there are nine prediction head in the paper's figure 3, but only 6 in the yaml files.

**Comment by wgqhandsome** [comment]
> there is 3+3+3, 9 heads in fig, but just 6 heads in yaml, which make me confused, is there any details i missed? and as shown in yolov9-e.yaml, the relationship between main branch and aux branch is different with fig in paper,Looking forward to your relpy! @WongKinYiu @sanha9999 ![4e94bb69ec3372213726b0f69a370cde](https://private-user-images.githubusercontent.com/50388099/311196550-563fc7b4-6ccb-42bf-af40-4a0e3c420efe.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk5ODQyNzcsIm5iZiI6MTcwOTk4Mzk3NywicGF0aCI6Ii81MDM4ODA5OS8zMTExOTY1NTAtNTYzZmM3YjQtNmNjYi00MmJmLWFmNDAtNGEwZTNjNDIwZWZlLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA5VDExMzI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkMjMzYjgyMDY1ZTQxMWM4YTUyOWQzYjZkNTdiZjM2OTNhYjVlNzFkYTJlNTQ3YTY2Y2E2OGQ4MmYyYzI1ZDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TIPOo_uVnBg1lvJov_3Fy1W6RvgOHKk4T2jbTpjWoWo)

I have the same question. There are only six heads in the model in the yaml file of yolo9-e. I am confused. If anyone can understand, please tell me.

**Comment by WongKinYiu** [comment]
You could take a look Table 4, if you have aux branches on both backbone and neck, you will have 3+3+3 heads.
And you could use train_triple.py to train the model.

**Comment by ccblublu** [comment]
If I want to have 9 heads, I need to select the corresponding YAML file. Based on the open-source code, I need to add the corresponding layers as inputs for the prediction head. From my observation, the main differences between train_*.py s are in the loss function and the selection of the detection head. Am I correct in my understanding? @WongKinYiu

**Comment by ccblublu** [comment]
> I have this question too. I already have found the multi-level auxiliary branch in yaml files, but, I have not find the Auxiliary Reversible Branch. And, there are nine prediction heads in the paper's figure 3, but only 6 in the yaml files. Why is this?

i think the Reversible structure is multi-level connections and fuse

**Comment by masc-it** [comment]
@WongKinYiu Could we have a diagram or explanation of what an auxiliary branch does in terms of operations/transformations, maybe some links to the code as well? 

It's clear that it serves the purpose of preserving the input-target relation throughout the layers, in a parallel branch, but said like that is kinda black magic to me.

The same applies to the multi level auxiliary branch and predictions heads: how they are plugged in the rest of the network?

**Comment by JaneM1222** [comment]
Which part of the source code specifically implements the operation of the reversible auxiliary branch? There are still doubts about the implementation of this part of the function. Can you explain the specific code?  @WongKinYiu

**Comment by WongKinYiu** [comment]
> Fig中有3+3+3，9個頭，但yaml中只有6個頭，這讓我很困惑，有什麼我錯過的細節嗎？ 並且如yolov9-e.yaml所示，main分支和aux分支的關係與紙上的圖不同，期待你的relpy！

![image](https://github.com/WongKinYiu/yolov9/assets/12152972/b02675d5-1898-4238-a22a-4dd79043fe37)

The main branch of yolov9-e contains reversible architecture.
The architecture of yolov9-e is modify from [dynamic-yolov7](https://openaccess.thecvf.com/content/CVPR2023/papers/Lin_DynamicDet_A_Unified_Dynamic_Architecture_for_Object_Detection_CVPR_2023_paper.pdf).

**Comment by WongKinYiu** [comment]
> Which part of the source code specifically implements the operation of the reversible auxiliary branch? There are still doubts about the implementation of this part of the function. Can you explain the specific code?

https://github.com/WongKinYiu/yolov9/blob/main/models/detect/yolov9-c.yaml#L81-L116

**Comment by masc-it** [comment]
A lack of proper explanation and understanding won't let this model have the attention it deserves. Please, consider writing a technical post or something where you explain in a more detailed view where, how and why things are placed in a certain way and why they work. Pointing to a Yaml it s of no help. 

Also, the yolov5 codebase you're based on builds the model arch at runtime and that doesn't help to visualize the flow. A class with all the layers would help as well.

I'm a scientist like you, but please be more careful to code quality and proper documentation. That stuff makes the difference. kudos for your next works

**Comment by WongKinYiu** [comment]
1. Original deep networks will lost information when feed forward.

![image](https://github.com/WongKinYiu/yolov9/assets/12152972/609c5460-8e45-40c8-84d7-1ce92a5f4dc9)

2. Especially loss the information for making projection from data to target.

![image](https://github.com/WongKinYiu/yolov9/assets/12152972/743c91c4-66e1-4ce7-aa4a-2678f787422e)

3. We can visualize that the information lost make we can not find correct relation to project data to target.

![image](https://github.com/WongKinYiu/yolov9/assets/12152972/35c6033c-b505-4af6-8616-8d6f87e7ccdd)

4. And modern networks could maintain more reliable information.

![image](https://github.com/WongKinYiu/yolov9/assets/12152972/35d62689-acbc-481b-9279-cdaee47efcc6)

5. Theoretically, reversible architecture could make deeper networks to maintain most of information of data.

![image](https://github.com/WongKinYiu/yolov9/assets/12152972/55d11c48-1208-49f4-a7ab-e9cb8db2c68a)

6. We also show the evidence that reversible branch make more trustworthy relation between data and target.

![image](https://github.com/WongKinYiu/yolov9/assets/12152972/e31c2620-a332-40d1-b7cd-3d7f9548a640)

7. And the corresponding part of reversible architecture is here.

https://github.com/WongKinYiu/yolov9/blob/main/models/detect/yolov9-c.yaml#L81-L116

**Comment by masc-it** [comment]
thanks, I've a couple of questions: 

1. How did you find that gelan actually satisfied the "reversible" property? Just by looking at feature maps? What's behind elan in general that makes it better than CSP and alikes ? 
2. What's the role of CBLinear from CBNet?
3. Have you done ablations of some sort?

**Comment by WongKinYiu** [comment]
no, gelan does not satisfied the reversible property, but cbnet and revcol do.
for analysis of elan and csp, please take a look [elan paper](https://jise.iis.sinica.edu.tw/JISESearch/pages/View/PaperView.jsf;jsessionid=73345fd66ee22916fc8edffe01b7?keyId=190_2662).

cblinear is just a set of linear layers to make different pyramidal feature maps to have same channel.
the main part which contains reversible property is cbfuse which composite higher level features of the first backbone into lower level of features of the second backbone.

we have examine different sorts composite ways in cbnet, the best performance is conduct by fully dhlc and fcc.
however, because we may add reversible auxiliary branch on different position of gelan-pan architecture.
for fair comparison, we only make dhlc composition on p3/p4/p5 levels.
the related ablations are shown in table 4.

**Comment by Muyyong** [comment]
> no, gelan does not satisfied the reversible property, but cbnet and revcol do. for analysis of elan and csp, please take a look [elan paper](https://jise.iis.sinica.edu.tw/JISESearch/pages/View/PaperView.jsf;jsessionid=73345fd66ee22916fc8edffe01b7?keyId=190_2662).
> 
> cblinear is just a set of linear layers to make different pyramidal feature maps to have same channel. the main part which contains reversible property is cbfuse which composite higher level features of the first backbone into lower level of features of the second backbone.
> 
> we have examine different sorts composite ways in cbnet, the best performance is conduct by fully dhlc and fcc. however, because we may add reversible auxiliary branch on different position of gelan-pan architecture. for fair comparison, we only make dhlc composition on p3/p4/p5 levels. the related ablations are shown in table 4.

Can I interpret the nearest sample and Add operation in CBFuse as a reversible structure ? or does it mean that the structure of fusing high-level information of the first backbone into the second backbone is reversible ? In other words,  if replacing CBFuse with Concat still a reversible structure?

**Comment by WongKinYiu** [comment]
Concat is OK
